File: comparaisonRemplissage.py
### Ajouté par l'API avant l'envoie à judge0
#JEPSEN

import cv2 as cv

from bibliothequePython.bib import Matrice, Pixel, Opencv, Exercice



def testAlgo(nameExercice, resultat, nbMatriceResult):
    opencv = Opencv(nameExercice)
    opencv.setNumberImageResultat(nbMatriceResult)
    opencv.getNumberImage()
    opencv.extractImage()
    opencv.initSizeImage()

    exercice = Exercice(resultat, nameExercice)
    listPatternInit = opencv.initExercice(opencv.sizeImage)
    print("test")
    print(opencv.sizeImage)

    print("pattern")
    for patern in listPatternInit:
        print(patern)



    solution_user = doExercice(listPatternInit, opencv.sizeImage)
    return exercice.assertRes(solution_user, resultat)




### FIN Ajouté par l'API avant l'envoie à judge0

#Consigne: Vous disposez d'une liste d'image.
#Implémentez la fonction doExercice afin de créer un algoritme capable de trouver le pattern remplie (dont tous les pixels intèriueurs sont noir)
#
#
#Donnée: Liste d'images a traiter => listMatrice

#Réponse: vous devez retourner l'id de limage dans la liste (0, 1, 2 ...) correspondant à une forme pleine


### Algo crée par l'utilisateur




def doExercice(listPatternInit,size_matrice):


    res1 = remplissageBalayage(listPatternInit, size_matrice)

    return res1





def remplissageBalayage(listPattern, size_matrice):
    idSolution = 0

    for pattern in listPattern:

        ligne = 0
        colonne = 0
        echec = False

        while ligne < size_matrice-1 and not echec:

            cpt_bordure = 0
            inForm = False

            while colonne < size_matrice-1 and not echec:

                if cpt_bordure > 2 and cpt_bordure != 0:
                    echec = True


                if pixelIsNotNull(colonne, ligne, pattern):


                    if not inForm:

                        inForm = True
                        cpt_bordure += 1

                else:

                    if inForm:

                        inForm = False
                        cpt_bordure += 1



                colonne = colonne + 1

            colonne = 0
            ligne = ligne + 1




        if not echec:
            return idSolution


        idSolution +=1


    return 0


def decalePixel(x, y, pattern, listPixelForme, size_matrice):

        if x <= size_matrice-1 and y<= size_matrice-1:

            if (x, y) not in listPixelForme :

                if pixelIsNotNull(x,y,pattern):


                    listPixelForme.append((x, y))

                    return decalePixel(x+1,y,pattern, listPixelForme, size_matrice) and \
                    decalePixel(x,y+1,pattern, listPixelForme, size_matrice) and \
                    decalePixel(x-1,y,pattern, listPixelForme, size_matrice) and \
                    decalePixel(x,y-1,pattern, listPixelForme, size_matrice)

                else:
                    return False

            else:
                return True

        else :
            return False


def pixelIsNotNull(x, y, matriceSource):

    pixelNone = Pixel(x, x, (255, 255, 255))

    return not matriceSource.getPixel(x, y).compare(pixelNone)




### FIN  Algo crée par l'utilisateur





if __name__ == '__main__':
    nameExercice = "comparaisonRemplissage"
    resultat = 2
    nbMatriceResult = 0
    print(testAlgo(nameExercice, resultat, nbMatriceResult))

File: cpp/comparaisonRemplissage.cpp

#include <stdio.h> 
#include <opencv2/highgui.hpp>
#include <string>
#include <iostream>
#include "./bibliothequeCpp/Opencv.h"
#include "./bibliothequeCpp/Exercice.h"
#include "./bibliothequeCpp/Matrice.h"
#include "./bibliothequeCpp/Line.h"
#include "./bibliothequeCpp/Pixel.h"
// g++ comparaisonMatrice.cpp -o comparaisonMatrice `pkg-config --cflags --libs opencv4`
using namespace std;







bool pixelIsNull(int x, int y, Matrice matriceSource){

    int color[3] = {255,255,255};
    Pixel pixelNone = Pixel(x, y, color);

    return !matriceSource.getPixel(x, y).compare(pixelNone);
}



int remplissageBalayage(vector<Matrice> listPattern, int size_matrice){
    int idSolution = 0;

    for(unsigned int i = 0; i < listPattern.size(); i++){
            Matrice pattern = listPattern[i];
            pattern.toStringPixel();
            bool echec = false;
            int ligne = 0;
            int colonne = 0;

        while (ligne < size_matrice-1 && !echec){

                int cpt_bordure = 0;
                bool inForm = false;


            while (colonne < size_matrice-1 && !echec){
  

                if (cpt_bordure > 2 && cpt_bordure != 0){
                    echec = true;
                }

                if (pixelIsNull(colonne, ligne, pattern)){

                    if (!inForm){
                        inForm = true;
                        cpt_bordure += 1;
                    }

                }else{

                    if (inForm){
                        inForm = false;
                        cpt_bordure += 1;
                    }
                }



                colonne = colonne + 1;

            }
            colonne = 0;
            ligne = ligne + 1;
        }



        if (!echec){
            return idSolution;
        }
        idSolution++;
        }

    return -1;
    }



int doExercice(vector<Matrice> listPatternInit,int size_matrice){

    int res1 = remplissageBalayage(listPatternInit, size_matrice);
    return res1;
}



string testAlgo(string nameExercice, int resultat, int nbMatriceResult){
    Opencv opencv = Opencv(nameExercice);
    opencv.setNumberImageResultat(nbMatriceResult);
    opencv.getNumberImage();
    opencv.extractImage();
    opencv.initSizeImage();

    Exercice exercice = Exercice(resultat, nameExercice);
    vector<Matrice> listPatternInit = opencv.initExercice();

    int solution_user = doExercice(listPatternInit, opencv.getSizeImage());
    cout << "SOLUTION" << endl;
    cout << solution_user << endl;

    return exercice.assertRes(solution_user, resultat);
}


int main(int argc, char *argv[])
{
    string nameExercice = "comparaisonRemplissage";
    int resultat = 2;
    int nbMatriceResult = 0;

    cout << testAlgo(nameExercice, resultat, nbMatriceResult) << endl;

}